requestmanager/reconciledloader/injest.go

```diff
@@ -0,0 +1,43 @@
+package reconciledloader
+
+import (
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-graphsync"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// IngestResponse ingests new remote items into the reconciled loader
+func (rl *ReconciledLoader) IngestResponse(md graphsync.LinkMetadata, traceLink trace.Link, blocks map[cid.Cid][]byte) {
+	if md.Length() == 0 {
+		return
+	}
+	duplicates := make(map[cid.Cid]struct{}, md.Length())
+	items := make([]*remotedLinkedItem, 0, md.Length())
+	md.Iterate(func(link cid.Cid, action graphsync.LinkAction) {
+		newItem := newRemote()
+		newItem.link = link
+		newItem.action = action
+		_, isDuplicate := duplicates[link]
+		if action == graphsync.LinkActionPresent && !isDuplicate {
+			duplicates[link] = struct{}{}
+			newItem.block = blocks[link]
+		}
```

Comment:
Hey, some states that I'm wondering about:

1. what is this going to do during processing in the case of a duplicate? it'll get an item with a "present" action but no block, is that OK, will it just skip over it as if it's not present?
2. what about the case where we have a duplicate metadata but with different actions? they'll both get queued up, is that OK?
3. what about duplicate not-present blocks? is that OK?

Reply:
maybe the parent question to all of these is - why are we concerned about duplicates at this particular point?

Reply:
so:

1. Currently, we haven't implemented the DuplicateNotSent in the metadata -- a duplicate is simply a Present where there is no block in the message. Currently, the remote once it has sent a block in one message will drop it from future messages for the same request. So we already handle the case where the action is Present but there the block == nil inside the reconciled loader.

2. This code processes metadata+blocks for a single message. If two metadatas with Action = Present for the same request in a single message reference the same block AND the block is present in the message, without the duplicates check here, we'd end up attaching the block to both metadatas. Tn memory reference would end up held until we drop the second metadata, and the block would be saved twice. Since we already have the code in #1 to deal with block == nil, this seemed like a simple high value optimization worth pursuing (thought possibly premature?).

3. Note there's a potential duplication in memory that will happen anyway -- if two metadatas with Action = Present for DIFFERENT requests in a single message reference the same block AND the block is present in the message, the block will end up in both requests queues, remaining in memory till both requests have processed it and possibly getting saved to the same store twice. At the same time, this is important to have in case the requests are writing to different stores -- you actually may need to save twice.